runtime/shared_common/OSCache.cpp

```diff
@@ -739,20 +741,55 @@ SH_OSCache::getAllCacheStatistics(J9JavaVM* vm, const char* ctrlDirName, UDATA g
 
 	/* Walk all the cache files in the j9shmem dir and (optionally) the persistent cacheDir */
 	do {
-		if (getCacheStatistics(vm, ctrlDirName, (const char*)nameWithVGen, groupPerm, localVerboseFlags, j2seVersion, &tempInfo, reason) != -1) {
+		J9Pool* lowerLayerList = NULL;
+		J9Pool** lowerLayerListPtr = &lowerLayerList;
+		bool isToplayer = isTopLayerCache(vm, ctrlDirName, nameWithVGen);
+		bool getLowerLayerStats = false;
+
+		if (getCacheStatistics(vm, ctrlDirName, (const char*)nameWithVGen, groupPerm, localVerboseFlags, j2seVersion, &tempInfo, reason, getLowerLayerStats, isToplayer, lowerLayerListPtr, NULL) != -1) {
 			if (includeOldGenerations || !tempInfo.isCompatible || (getCurrentCacheGen() == tempInfo.generation)) {
 				if (tempInfo.isCompatible) {
 					if (!ignoreCompatible) {
-						newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
-						memcpy(newElement, &tempInfo, sizeof(SH_OSCache_Info));
+						if (isToplayer) {
+							/* nameWithVGen is a compatible top layer cache, lower layer stats are in lowerLayerList*/
+							newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
+							memcpy(newElement, &tempInfo, sizeof(SH_OSCache_Info));
+							if (NULL != lowerLayerList) {
+								if (0 < pool_numElements(lowerLayerList)) {
+									pool_state poolState;
+									SH_OSCache_Info* anElement = (SH_OSCache_Info*)pool_startDo(lowerLayerList, &poolState);
+									do {
+										newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
+										if (NULL == newElement) {
+											break;
+										}
+										memcpy(newElement, anElement, sizeof(SH_OSCache_Info));
+										++cntr;
+									} while ((anElement = (SH_OSCache_Info*)pool_nextDo(&poolState)) != NULL);
+								}
+								pool_kill(lowerLayerList);
+							}
+						} else {
+							if (SHR_STATS_REASON_ITERATE != reason) {
+								newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
+								if (NULL != newElement) {
```

Comment:
Seems weird to just ignore the failure.

Reply:
Changed to break in failure.

Reply:
I still don't see how that doesn't ignore the failure.

Reply:
I was thinking if pool_newElement() failed, we can break but still return existing SH_OSCache_Info(s) in cacheList. Even the cacheList is incomplete, it is better than returning NULL. Do you want a trace point to log the failure ? I see there is an exit tracepoint inside pool_newElement() logging the return value.

Reply:
A tracepoint would be good. It would be hard to associate the pool_newElement() tracepoint with the shared classes cache. Why is a partial cacheList better than failing? It means the result is incomplete, but there is no indication to the user that it is incomplete. If it fails there is either a bug we need to resolve, or we are out of native memory which is going to be fatal somewhere else anyway.

Reply:
I've added Trc_SHR_OSC_getAllCacheStatistics_pool_newElement_failed() to log the failure. 
For the cacheList in case of pool_newElement() failure, probably I should say an incomplete list is not worse than NULL. But I don't have a strong preference between these two behaviours.

Reply:
My preference is to return NULL, which is a stronger indication that something is wrong and more likely to fail testing and result in a fix if there is a bug. In theory this should never happen anyway.

Reply:
Changed to return NULL if pool_newElement() fails.